File: anna/normalize.js
// anna/normalize.js
// ─────────────────────────────────────────────────────────────────────────────
// Turning messy input into a structured DriverProfile.
//
// Two jobs:
//   1. normalizeDriver()      — map a raw lead (varied ad-source fields, free
//                               text) onto Anna's structured DriverProfile.
//   2. extractFromDocument()  — read a license / medical card / cert image|PDF
//                               and pull the fields needed to auto-fill the
//                               application (Stage 2).
//
// Both prefer Claude when ANTHROPIC_API_KEY is set, and fall back to a
// deterministic passthrough so the module is usable (and unit-testable) without
// an API key. Every LLM-extracted field carries a confidence; low-confidence
// fields are surfaced for human/driver confirmation rather than trusted blindly.
// ─────────────────────────────────────────────────────────────────────────────

import { callClaudeJSON, documentBlock, MODELS, annaConfigured } from './claude.js';

// The canonical shape the matcher consumes. Documented here as the contract.
export const DRIVER_SHAPE = {
  name: 'string', age: 'number', phone: 'string', email: 'string',
  cdl: { class: 'A|B|C', endorsements: ['H', 'N', 'T', 'X', 'P', 'S'], experienceYears: 'number', type: 'company|owner-operator', expiresInDays: 'number' },
  mvr: { movingViolations: 'number', accidents: 'number', dui: 'number' },
  psp: { crashes: 'number', oosInspections: 'number' },
  insurance: { autoLiability: 'number', hasCargo: 'boolean', cargo: 'number' },
};

const num = (x) => (Number.isFinite(Number(x)) ? Number(x) : undefined);

/**
 * Normalize a raw lead into a DriverProfile.
 * @param {Object} lead  Arbitrary lead payload (FB/TikTok/form fields, free text).
 * @param {Object} [opts] { apiKey, model }
 * @returns {Promise<{ profile, confidence, source }>}
 */
export async function normalizeDriver(lead = {}, opts = {}) {
  if (annaConfigured(opts.apiKey)) {
    try {
      const out = await callClaudeJSON({
        apiKey: opts.apiKey,
        model: opts.model || MODELS.fast,
        maxTokens: 1024,
        system:
          'You normalize raw truck-driver lead data into a strict JSON DriverProfile for an FMCSA driver-qualification system. ' +
          'Infer numeric fields from free text (e.g. "drove for Swift 2018-2021" => experienceYears 3). ' +
          'Never invent data: if a field is unknown, omit it. Return ONLY JSON.',
        messages: [{
          role: 'user',
          content:
            `Raw lead:\n${JSON.stringify(lead, null, 2)}\n\n` +
            'Return JSON: { "profile": <DriverProfile>, "confidence": { "<dot.path>": 0..1 } }. ' +
            `DriverProfile shape: ${JSON.stringify(DRIVER_SHAPE)}.`,
        }],
      });
      return { profile: coerce(out.profile || {}), confidence: out.confidence || {}, source: 'ai' };
    } catch (e) {
      // Fall through to deterministic mapping; never block intake on an API error.
    }
  }
  return { profile: deterministicMap(lead), confidence: {}, source: 'heuristic' };
}

/**
 * Extract structured fields from a document for application auto-fill.
 * @param {Object} p { dataUrl|base64+mediaType, docType, apiKey, model }
 * @returns {Promise<{ fields, confidence, warnings }>}
 */
export async function extractFromDocument(p = {}) {
  if (!annaConfigured(p.apiKey)) throw new Error('Document scanning requires ANTHROPIC_API_KEY');
  const docType = p.docType || 'document';
  const out = await callClaudeJSON({
    apiKey: p.apiKey,
    model: p.model || MODELS.fast,
    maxTokens: 1024,
    system:
      'You read US commercial driver documents (CDL, DOT medical card, certificates) and extract fields for an application. ' +
      'Report a confidence 0..1 per field. Flag expired documents and unreadable fields in "warnings". Return ONLY JSON.',
    messages: [{
      role: 'user',
      content: [
        documentBlock(p),
        { type: 'text', text:
          `This is a "${docType}". Extract relevant fields (e.g. for a CDL: fullName, licenseNumber, state, class, endorsements[], issueDate, expirationDate). ` +
          'Return JSON: { "fields": {...}, "confidence": {"<field>":0..1}, "warnings": ["..."] }.' },
      ],
    }],
  });
  return { fields: out.fields || {}, confidence: out.confidence || {}, warnings: out.warnings || [] };
}

/** Fields that should be human-confirmed (below threshold) before auto-advancing. */
export function lowConfidenceFields(confidence = {}, threshold = 0.75) {
  return Object.entries(confidence).filter(([, c]) => Number(c) < threshold).map(([k]) => k);
}

// ── Deterministic fallback: copy known fields straight through, coerce types. ──
function deterministicMap(lead) {
  // Accept either an already-nested profile or a flat lead with matching keys.
  if (lead.cdl || lead.mvr || lead.psp || lead.insurance) return coerce(lead);
  return coerce({
    name: lead.name, age: lead.age, phone: lead.phone, email: lead.email,
    cdl: { class: lead.cdlClass, endorsements: lead.endorsements, experienceYears: lead.experienceYears, type: lead.driverType, expiresInDays: lead.cdlExpiresInDays },
    mvr: { movingViolations: lead.movingViolations, accidents: lead.accidents, dui: lead.dui },
    psp: { crashes: lead.crashes, oosInspections: lead.oosInspections },
    insurance: { autoLiability: lead.autoLiability, hasCargo: lead.hasCargo, cargo: lead.cargo },
  });
}

// Coerce numeric fields and drop empties so the matcher sees clean data.
function coerce(p = {}) {
  const out = { ...p };
  if (p.age != null) out.age = num(p.age);
  out.cdl = clean({ ...p.cdl, experienceYears: num(p.cdl?.experienceYears), expiresInDays: num(p.cdl?.expiresInDays), class: p.cdl?.class ? String(p.cdl.class).toUpperCase() : undefined });
  out.mvr = clean({ movingViolations: num(p.mvr?.movingViolations), accidents: num(p.mvr?.accidents), dui: num(p.mvr?.dui) });
  out.psp = clean({ crashes: num(p.psp?.crashes), oosInspections: num(p.psp?.oosInspections) });
  out.insurance = clean({ autoLiability: num(p.insurance?.autoLiability), hasCargo: p.insurance?.hasCargo, cargo: num(p.insurance?.cargo) });
  return out;
}

function clean(o) {
  const out = {};
  for (const [k, v] of Object.entries(o)) if (v !== undefined && v !== null && v !== '') out[k] = v;
  return out;
}

File: backend/app/anna/normalize.py
"""Turning messy input into a structured DriverProfile (Claude + fallback)."""
import json

from .claude import MODELS, anna_configured, call_claude_json, document_block

DRIVER_SHAPE = {
    "name": "string", "age": "number", "phone": "string", "email": "string",
    "cdl": {"class": "A|B|C", "endorsements": ["H", "N", "T", "X", "P", "S"], "experienceYears": "number", "type": "company|owner-operator", "expiresInDays": "number"},
    "mvr": {"movingViolations": "number", "accidents": "number", "dui": "number"},
    "psp": {"crashes": "number", "oosInspections": "number"},
    "insurance": {"autoLiability": "number", "hasCargo": "boolean", "cargo": "number"},
}


def _num(x):
    try:
        n = float(x)
        if n != n:
            return None
        return int(n) if n == int(n) else n
    except (TypeError, ValueError):
        return None


async def normalize_driver(lead: dict | None = None, opts: dict | None = None) -> dict:
    lead = lead or {}
    opts = opts or {}
    if anna_configured(opts.get("apiKey")):
        try:
            out = await call_claude_json(
                api_key=opts.get("apiKey"),
                model=opts.get("model") or MODELS["fast"],
                max_tokens=1024,
                system=(
                    "You normalize raw truck-driver lead data into a strict JSON DriverProfile for an FMCSA driver-qualification system. "
                    'Infer numeric fields from free text (e.g. "drove for Swift 2018-2021" => experienceYears 3). '
                    "Never invent data: if a field is unknown, omit it. Return ONLY JSON."
                ),
                messages=[{
                    "role": "user",
                    "content": (
                        f"Raw lead:\n{json.dumps(lead, indent=2)}\n\n"
                        'Return JSON: { "profile": <DriverProfile>, "confidence": { "<dot.path>": 0..1 } }. '
                        f"DriverProfile shape: {json.dumps(DRIVER_SHAPE)}."
                    ),
                }],
            )
            return {"profile": _coerce(out.get("profile") or {}), "confidence": out.get("confidence") or {}, "source": "ai"}
        except Exception:  # noqa: BLE001
            pass
    return {"profile": _deterministic_map(lead), "confidence": {}, "source": "heuristic"}


async def extract_from_document(p: dict | None = None) -> dict:
    p = p or {}
    if not anna_configured(p.get("apiKey")):
        raise RuntimeError("Document scanning requires ANTHROPIC_API_KEY")
    doc_type = p.get("docType") or "document"
    out = await call_claude_json(
        api_key=p.get("apiKey"),
        model=p.get("model") or MODELS["fast"],
        max_tokens=1024,
        system=(
            "You read US commercial driver documents (CDL, DOT medical card, certificates) and extract fields for an application. "
            'Report a confidence 0..1 per field. Flag expired documents and unreadable fields in "warnings". Return ONLY JSON.'
        ),
        messages=[{
            "role": "user",
            "content": [
                document_block(data_url=p.get("dataUrl"), base64_data=p.get("base64"), media_type=p.get("mediaType")),
                {"type": "text", "text": (
                    f'This is a "{doc_type}". Extract relevant fields (e.g. for a CDL: fullName, licenseNumber, state, class, endorsements[], issueDate, expirationDate). '
                    'Return JSON: { "fields": {...}, "confidence": {"<field>":0..1}, "warnings": ["..."] }.'
                )},
            ],
        }],
    )
    return {"fields": out.get("fields") or {}, "confidence": out.get("confidence") or {}, "warnings": out.get("warnings") or []}


def low_confidence_fields(confidence: dict | None = None, threshold: float = 0.75) -> list:
    confidence = confidence or {}
    return [k for k, c in confidence.items() if (_num(c) or 0) < threshold]


def _deterministic_map(lead: dict) -> dict:
    if lead.get("cdl") or lead.get("mvr") or lead.get("psp") or lead.get("insurance"):
        return _coerce(lead)
    return _coerce({
        "name": lead.get("name"), "age": lead.get("age"), "phone": lead.get("phone"), "email": lead.get("email"),
        "cdl": {"class": lead.get("cdlClass"), "endorsements": lead.get("endorsements"), "experienceYears": lead.get("experienceYears"), "type": lead.get("driverType"), "expiresInDays": lead.get("cdlExpiresInDays")},
        "mvr": {"movingViolations": lead.get("movingViolations"), "accidents": lead.get("accidents"), "dui": lead.get("dui")},
        "psp": {"crashes": lead.get("crashes"), "oosInspections": lead.get("oosInspections")},
        "insurance": {"autoLiability": lead.get("autoLiability"), "hasCargo": lead.get("hasCargo"), "cargo": lead.get("cargo")},
    })


def _clean(o: dict) -> dict:
    return {k: v for k, v in o.items() if v is not None and v != ""}


def _coerce(p: dict | None = None) -> dict:
    p = p or {}
    out = {**p}
    if p.get("age") is not None:
        out["age"] = _num(p.get("age"))
    cdl = p.get("cdl") or {}
    out["cdl"] = _clean({
        **cdl,
        "experienceYears": _num(cdl.get("experienceYears")),
        "expiresInDays": _num(cdl.get("expiresInDays")),
        "class": str(cdl.get("class")).upper() if cdl.get("class") else None,
    })
    mvr = p.get("mvr") or {}
    out["mvr"] = _clean({"movingViolations": _num(mvr.get("movingViolations")), "accidents": _num(mvr.get("accidents")), "dui": _num(mvr.get("dui"))})
    psp = p.get("psp") or {}
    out["psp"] = _clean({"crashes": _num(psp.get("crashes")), "oosInspections": _num(psp.get("oosInspections"))})
    ins = p.get("insurance") or {}
    out["insurance"] = _clean({"autoLiability": _num(ins.get("autoLiability")), "hasCargo": ins.get("hasCargo"), "cargo": _num(ins.get("cargo"))})
    return out
